DAND-43-Add scripts for monorepo automation

OllaAshour opened the pull request:
**What does this PR do?**
-Add script to install any missing/required tools and dependencies
-Add gem files to maintain same environment for all developers
-Clean project
-Run pod install/update
-Test project
-Run OCLint checks
-Add clang-format to format objective c code.

**To Use**

- Please run check.sh to make sure all dependencies are installed either within the scripts directory or from parent directory using `./start.sh sh check.sh`

- Then run `./start.sh sh update.sh` to install and update pods.

- Then run `./start.sh sh test.sh` to make sure all tests have passed
 
- Then run `./start.sh sh lint.sh xxxx` to lint, and to clean before building and linting.
xxxx is target name. `./start.sh sh lint.sh Segment-Firebase`
and to format, run`./start.sh sh format.sh`

**Associated JIRA Tickets:**
https://segmentcontractors.atlassian.net/browse/DAND-47

OllaAshour committed:
DAND-43-Add gem file for environment consistencies

(cherry picked from commit 1a37cc186b5eb0c09026bda7bbeb2901f00183e3)

OllaAshour committed:
DAND-43-Add script for gem dependencies, pod install and update

(cherry picked from commit a9e97590f1735ce76d2d973deadb99dbbfa9e10d)

OllaAshour committed:
DAND-43- Separate scripts and add scheme checks

OllaAshour committed:
DAND-43-Add clang format rules, and for formatting objective c code.

(cherry picked from commit fdddd74376baea077ffa5a9dffa50f83ce3cd185)

OllaAshour committed:
DAND-43-Update scripts to be more user friendly and print status.

(cherry picked from commit 2ac07da21aa55fd63ec962545f3cd32da4064999)

OllaAshour committed:
DAND-43-Added folder for scripts and renamed scipts. Also added formating using clang-format to lint script

(cherry picked from commit cb64570a25e24bfc4959d4e7e75b8128928aa691)

OllaAshour committed:
DAND-43-Clean up .check script and gem file

(cherry picked from commit 9b614ae95cea3af0541cec3a12ac89577a9a14b2)

OllaAshour committed:
DAND-43-Update Releasing.md guide

(cherry picked from commit c9db69c881e4062b72cadcb1f41eff38efcedf2d)

OllaAshour committed:
DAND-43-update chmod on test.sh and update.sh scripts

(cherry picked from commit a925117ef56e436a5411841a1c28ae0cabd9efc1)

OllaAshour committed:
DAND-43-add start.sh to allow scripts to be run from parent directory

(cherry picked from commit 6f4925de2f78afb6f84e3040faa7f2b004aba7a8)

OllaAshour committed:
DAND-43-remove code signing team from targets

(cherry picked from commit 083fa1442920dd5d3be7f315c695dcc11fc955cb)

OllaAshour committed:
DAND-43-Update OCLint exclude checks to include Pods directory

(cherry picked from commit cb526cff46508e6d45194ccfa4a3f4a28511b1e3)

briemcnally commented:
@OllaAshour So I ran
1.`.start.sh`
2. `update.sh`
3. `check.sh`

Those all look good. The `./lint.sh` is printing that it generated ton of errors for me. You mentioned it is related to path in podspec. Can you update this so we can observer correct behavior or the linter? 

Also the `./test.sh` command is not running the tests still on this PR. From root directory I get the error: 
```
xcodebuild: error: The directory /Users/brienne.mcnally/dev/src/github.com/segmentio does not contain an Xcode project, workspace or package.
```
Within the Scripts directory I get this error for `.test.sh`:
```
note: Using new build system
note: Planning build
note: Constructing build description
error: Signing for "Segment-FacebookTests" requires a development team. Select a development team in the Signing & Capabilities editor. (in target 'Segment-FacebookTests' from project 'Segment-Facebook')


Test session results, code coverage, and logs:
	/Users/brienne.mcnally/Library/Developer/Xcode/DerivedData/iOS-Integrations-azkshnsjcihbcrcfkrqvixhoinrq/Logs/Test/Test-Segment-Facebook-2020.01.13_17-39-49--0800.xcresult

Testing failed:
	Signing for "Segment-FacebookTests" requires a development team. Select a development team in the Signing & Capabilities editor.
	Testing cancelled because the build failed.

** TEST FAILED **
```

OllaAshour committed:
DAND-43-Move formatting into a separate bash script

OllaAshour committed:
DAND-43-Run formatting against all code

OllaAshour committed:
DAND-43-Add empty release.sh script for future

OllaAshour committed:
DAND-43-Update Readme.md on how to format, lint and test code. Also, how to submit a PR and what is required before submitting a PR

OllaAshour committed:
DAND-43- Add tests checks for mixpanel and firebase targets

OllaAshour committed:
DAND-43-Update Migrating document to reflect changes

OllaAshour committed:
DAND-43- Update Mixpanel scheme

OllaAshour committed:
DAND-43-Add command line tool as requirement

briemcnally commented on README.md:
```diff
@@ -1,2 +1,131 @@
+4- OCLint
+5- Clang-format
+
+Alternatively, run .check.sh to install any missing local dependencies.
```
Can we add a comment this needs to be ran from within the Scripts directory?

OllaAshour replied:
I'll add a note here to check "Usage" on how to run the scripts.

briemcnally commented on Guides/Migrating.md:
```diff
@@ -40,7 +40,15 @@ Also update dependencies version to latest if required.
 And run pod install. If successful, running and testing will work correctly. 
 
-#### 9. In example project, add the newly added destination by setting its relative path.
+#### 9. For tests to compile successfully from command line, code signing for the test targets must be set. (It's enough that it's a personal team, Apple ID does not need to have purchased developer program)
```
How do we set code signing for the test targets? I am not familiar with this?

OllaAshour replied:
Select Target and Go to General, and then Choose Development team.
<img width="886" alt="Screen Shot 2020-01-16 at 2 41 44 AM" src="https://user-images.githubusercontent.com/4341880/72542729-4e189700-388d-11ea-98a4-062d54f516ae.png">
 select Test Target.

briemcnally commented on Guides/Migrating.md:
```diff
@@ -40,7 +40,15 @@ Also update dependencies version to latest if required.
-#### 9. In example project, add the newly added destination by setting its relative path.
+#### 9. For tests to compile successfully from command line, code signing for the test targets must be set. (It's enough that it's a personal team, Apple ID does not need to have purchased developer program)
+
+#### 10. It may happen that the pod install has created extra schemes in workspace, simply delete them, but (choose Remove reference) upon delete confirmation.
```
Can you provide more detail for users unrelated to Xcode projects where they are supposed to be findnig/deleting these extra schems in the workspace. I think if we can make these steps as explicit as possible the better. Ie. Steps like go here in Xcode, Produc  -> Scheme -> Edit Scheme and then explicitly state which "extra schemes" the user should be removing or should deem as extra.

briemcnally commented on Guides/Migrating.md:
```diff
@@ -40,7 +40,15 @@ Also update dependencies version to latest if required.
+
+#### 10. It may happen that the pod install has created extra schemes in workspace, simply delete them, but (choose Remove reference) upon delete confirmation.
+
+#### 11. It may also happen that the test targets have not been created, in that case, select scheme -> edit scheme ->  test -> +  and it will automatically create a test scheme.
```
Again once they navigate to test what specifically are they hitting? For me it seems it the test scheme is already enabled so I can't see the workflow for when its not automatically enabled. But can we be as specific as to say be on the `Info` tab and hit `+`. Then choose the test target to add to this scheme based off the integration which you just migrated. Again the more explicit here the better.

briemcnally commented on Guides/Migrating.md:
```diff
@@ -40,7 +40,15 @@ Also update dependencies version to latest if required.
+
+#### 11. It may also happen that the test targets have not been created, in that case, select scheme -> edit scheme ->  test -> +  and it will automatically create a test scheme.
+
+#### 12. It may also happen that after pod install, all pods are added as targets, in that case, simply select "Manage Schemes" and uncheck all the unrequited targets. (Extra step that will not affect any of the project functionality)
```
I'm not entirely sure what you mean by the `unerquited targets`?  Here is what my manage schemes looks like for this PR. Can you specify?
<img width="913" alt="Screen Shot 2020-01-15 at 9 35 44 AM" src="https://user-images.githubusercontent.com/31782219/72456925-706dc000-377a-11ea-8670-0355494086bf.png">

OllaAshour replied:
Here because I fixed it. But sometimes, you will see e.g "Analytics", "Expecta", are selected, which is not necessary.
We just need the 3 targets we use only.
OCLint is an aggregated target, to show OCLint errors in Xcode rather than command line. 
I'll add more info on that.

briemcnally commented on Guides/Releasing.md:
```diff
@@ -2,9 +2,10 @@
-#### 6. `pod trunk push xxxx`, where xxxx is the newley added integration podspec.
\ No newline at end of file
+##### 1. Update the path of `Source` in the pod spec to point to master branch and change version for next release version.
+##### 2. Run ./lint.sh and .test.ch to make sure no errors detected and all tests pass.
```
The `./lint.sh` command is properly building the project now, the previous PR was not, however the actual linter is printing a ton of errors to the terminal without specific detail. 

It looks like a hundred lines of this:
```
16 errors generated.
16 errors generated.
20 errors generated.
21 errors generated.
16 errors generated.
20 errors generated.
21 errors generated.
```

Then the actual linter outputs this with no meaningful error log for the error:
```
Linting pod spec

 -> Segment-Facebook-App-Events (1.0.4)
    - WARN  | source: Git sources should specify a tag.
    - WARN  | [iOS] license: Unable to find a license file
    - ERROR | [iOS] xcodebuild: Returned an unsuccessful exit code. You can use `--verbose` for more information.
    - NOTE  | xcodebuild:  note: Using new build system
    - NOTE  | [iOS] xcodebuild:  note: Planning build
    - NOTE  | [iOS] xcodebuild:  note: Constructing build description
    - NOTE  | [iOS] xcodebuild:  warning: Capabilities for Signing & Capabilities may not function correctly because its entitlements use a placeholder team ID. To resolve this, select a development team in the App editor. (in target 'App' from project 'App')
    - NOTE  | [iOS] xcodebuild:  warning: Skipping code signing because the target does not have an Info.plist file and one is not being generated automatically. (in target 'App' from project 'App')
    - NOTE  | [iOS] xcodebuild:  clang: error: linker command failed with exit code 1 (use -v to see invocation)
    - NOTE  | [iOS] xcodebuild:  warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 7.0, but the range of supported deployment target versions is 8.0 to 13.2.99. (in target 'Analytics' from project 'Pods')

[!] Segment-Facebook-App-Events did not pass validation, due to 1 error.
You can use the `--no-clean` option to inspect any issue.
```

briemcnally commented on README.md:
````diff
@@ -1,2 +1,131 @@
+We use xcode build tools to run tests. You can run all the tests using:
+```bash
+$ ./test 
+```
````
This still doesn't seem to be working for me: 
```

❌  error: Signing for "Segment-FacebookTests" requires a development team. Select a development team in the Signing & Capabilities editor. (in target 'Segment-FacebookTests' from project 'Segment-Facebook')


Testing failed:
	Signing for "Segment-FacebookTests" requires a development team. Select a development team in the Signing & Capabilities editor.
	Testing cancelled because the build failed.

** TEST FAILED **


❌  error: Signing for "Segment-MixpanelTests" requires a development team. Select a development team in the Signing & Capabilities editor. (in target 'Segment-MixpanelTests' from project 'Segment-Mixpanel')


Testing failed:
	Signing for "Segment-MixpanelTests" requires a development team. Select a development team in the Signing & Capabilities editor.
	Testing cancelled because the build failed.

** TEST FAILED **


❌  error: Signing for "Segment-FirebaseTests" requires a development team. Select a development team in the Signing & Capabilities editor. (in target 'Segment-FirebaseTests' from project 'Segment-Firebase')


Testing failed:
	Signing for "Segment-FirebaseTests" requires a development team. Select a development team in the Signing & Capabilities editor.
	Testing cancelled because the build failed.

** TEST FAILED **
```

briemcnally reviewed (changes requested):
Left some comments. Still seems like `./test.sh` and `./lint.sh` are still not working properly. 

If the steps in the `Migrating.md` are necessary to run the `.test.sh` we should also include them in them in the `READme`.  In general can we go into more explicit detail on the instructions.

OllaAshour commented on README.md:
````diff
@@ -1,2 +1,131 @@
+```
+For the purposes of the guide, we will assume, you are inside the Scripts directory.
+
+1- Please run check.sh to make sure all required dependencies have been installed on machine
````
@briemcnally I added in Usage how to run within Scripts directory or in root of project.

OllaAshour committed:
DAND-43-Add license files for facebook, firebase and mixpanel and changelog for each of the repos

OllaAshour committed:
DAND-43-Update pod file to set all of pod targets as minimum 9.0.
Also update pods

OllaAshour committed:
DAND-43-Update lint script to take Targets dynamically and also directory paths of destinations

OllaAshour committed:
DAND-43-Update podspec with source destinations

OllaAshour committed:
DAND-43-Update Simple Example pods

OllaAshour commented:
@briemcnally I have updated the lint script. I think now everything should work correctly. It was because I was including files that should not be linted.
I also updated the lint script to take target name as an argument, to allow linting of any destination.
Also updated README.md and Migrating with the comments

Do kindly test the scripts and hopefully everything works well.

To run lint please do the following `./start.sh sh lint.sh xxxx` where xxxx is target name, so e.g it would be `./start.sh sh lint.sh Segment-Firebase`

Note: Please add team for the test script to work correctly. Thanks

OllaAshour committed:
DAND-43-Update guides with comments
